cogs/utils/arg_converter.py

```diff
@@ -0,0 +1,150 @@
+# under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Canary is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Canary. If not, see <https://www.gnu.org/licenses/>.
+
+from discord.ext import commands
+
+
+class ArgConverter:
+    """
+    The ArgConverter object is used to pseudo-typecheck and cast
+    a list of arguments using the given converters, regardless of order.
+    Matches each converters at most once.
+
+    Arg: A dictionary where keys are a string representing the variable names,
+    and values are a tuple with the converter and the default value. If no
+    default value is given, the variable is required.
+
+    A converter is an object that tries to convert an argument to a certain
+    class. If it can't, it raises discord.ext.commands.BadArgument. It takes
+    as input a discord context and an argument.
+    There are many default Discord Converters, for example
+    commands.MemberConverter(). Otherwise, it is possible to make a converter
+    by inheriting commands.Converter. For example, here is a converter that
+    checks if an argument starts by "a" and returns the rest of the string
+    if so
+
+    class StartsByAConverter(commands.Converter):
+        async def convert(self, ctx, argument):
+            if not type(argument) is str:
+                raise commands.BadArgument("Unexpected type")
+            if argument[0] != a:
+                raise commands.BadArgument("Argument doesn't start with 'a'")
+            result = argument[1:]
+            return result
+
+    A converter could also simply check that an argument is of a certain type
+    or has certain attributes and return the argument itself with no
+    conversion, which would act like a pseudo-typecheck. It could also
+    call another converter, for example stripping a certain argument and
+    and trying to convert the remaining string to a discord member.
+
+    Example use:
+
+    from discord.ext.commands import MemberConverter, TextChannelConverter
+    class StartsByAConverter(commands.Converter):
+        (as above)
+    @commands.command()
+    async def example(self, ctx, *args):
+      arg_converter = ArgConverter({"member": (MemberConverter(), None),
+                                    "channel": (TextChannelConverter(), )
+                                    "string_A": (FromConverter(), "hi"))
+    try:
+        args_dict = await arg_converter.convert(ctx, args)
+    except commands.BadArgument as err:
+        await ctx.send(err)
+        return
+    await ctx.send(f"{str(member)}, {channel}, {string_A}")
```

Comment:
i think its supposed to be `await ctx.send(f"{str(args_dict['member'])}, {args_dict['channel']}, {args_dict['string_A']}")`

Reply:
Resolved in c6111c6